Fix the reflection primitive `getType` ignoring module parameters

WhatisRT opened the pull request:
`getType` currently ignores the context it is invoked from, returning the type a name would have outside of any module. This means that the type it returns might not actually be the type that a name has where it is currently invoked. This PR should give `getType` the expected behaviour in this scenario.

Alternatives to this would be to either make this behaviour controllable or to provide a primitive that returns the common module context with a name and to expect the user to manually adjust the type. I this those are both useful additions, but the behaviour implemented by this PR seems like a sensible default.

WhatisRT committed:
Fix the reflection primitive `getType` ignoring module parameters

WhatisRT committed:
Fix the reflection primitive `getType` ignoring module parameters

jespercockx commented:
Thank you for the PR! This is a very welcome change, however I am wondering whether we should still also have a primitive for the old behaviour, since sometimes you perhaps want to apply the definition to module parameters by hand.

nad commented:
Perhaps it would make sense to redesign the reflection machinery with more focus on "context" (see #3819 and #4598 for some other problems related to parametrised modules).

WhatisRT commented:
I was going to propose a redesign of some aspects of the reflection system at some point, and better context features would absolutely be part of that. I was going to bring that up whenever the next AIM happens. This particular one was just a low hanging fruit that's been causing me a lot of trouble.

If we had a primitive that returns the telescope of the current module, one could use that to implement a variant of `getType` with the old behaviour. I'd consider that behaviour edge-casey enough to drop it, but I'd also be willing to implement the primitive I just mentioned (either at some point, or as part of this PR if needed).

andreasabel commented:
@jespercockx :
> however I am wondering whether we should still also have a primitive for the old behaviour, since sometimes you perhaps want to apply the definition to module parameters by hand.

One can always switch to the top context with `inContext []`, I suppose, to get the old behavior.

andreasabel commented on CHANGELOG.md:
```diff
@@ -86,6 +86,10 @@ Language
 * A new constructor `pattErr : Pattern → ErrorPart` of `ErrorPart` for reflection
   is added.
 
+* The reflection primitive `getType` respects the module context it is
```
Equip this point with attribute `[breaking]` (exact form see other items).

andreasabel commented on CHANGELOG.md:
```diff
@@ -86,6 +86,10 @@ Language
 
+* The reflection primitive `getType` respects the module context it is
+  invoked from instead of returning the type a name would have outside
+  of all modules.
```
Mention the `inContext` trick to get the old behavior.

WhatisRT replied:
Turns out that that doesn't work, it seems like `instanciateDef` still does its job even if the context is reset. `inferType` has the same behaviour, which means that maybe `inContext` doesn't do what it should?

andreasabel replied:
Ok, maybe I was naive about it and going into the top context does not exit from the current module stack.

Maybe the original `getType` could be preserved under the name `getClosedType`?

Also, I wonder whether `getType` is now still in sync with `getDefinition`?  E.g., does `getDefinition` for a function instantiate the clauses to the current module parameters (if applicable)?

WhatisRT replied:
@jespercockx and I agreed now that we don't want to support this behaviour, since it's not something you could do outside of reflection anyway. I've also changed `getDefinition` to be in sync, which should fix #6200 then.

andreasabel replied:
> this behaviour

_This behavior_ is which behavior exactly?

WhatisRT replied:
The ability for `inContext` to go all the way to the top level. Jesper changed `inContext` to always stay relative to the context the `TC` computation started.

andreasabel reviewed (changes requested):
Please make the current docs more precise so that it tells you that the type will be relative to the current contexts. https://github.com/agda/agda/blob/514d6e53b6f22ef62d4090b870356ffd00e5dd03/doc/user-manual/language/reflection.lagda.rst?plain=1#L467-L468

andreasabel commented:
Lucas Escot discovered that there is an issue for this PR:
- #6200

jespercockx committed:
[ re #5858 ] Make `inContext` return to the original context of the macro call instead of the empty context

- Use `inOriginalContext` instead of `unsafeInTopContext` for `tcInContext`
- Better implementation of `inOriginalContext` using `escapeContext` instead of `unsafeModifyContext`

WhatisRT committed:
Also apply the change to `getDefinition` and add documentation

WhatisRT committed:
Document the changes to `inContext` as well

WhatisRT committed:
Fix the test

andreasabel commented:
@jespercockx Please merge and cherry-pick onto the release branch when ready.

andreasabel commented:
Cherry-picked to `release-2.6.3`.
